tools/castle-editor/code/framedesign.pas

```diff
@@ -1857,6 +1863,29 @@ procedure TDesignFrame.ControlsTreeDragOver(Sender, Source: TObject; X,
   ControlsTree.Invalidate; // force custom-drawn look redraw
 end;
 
+procedure TDesignFrame.ControlsTreeEditing(Sender: TObject; Node: TTreeNode;
+  var AllowEdit: Boolean);
+begin
+  { this event is fired when calling TCustomListView.CanEdit which itself is called in TCustomListView.ShowEditor
+    therefore this event preceeds initializing and showing of the editor }
+  //here we have to "restore" the pure name of the component (without class name) before starting edit
+  Node.Text := TComponent(ControlsTree.Selected.Data).name;
+end;
+
+procedure TDesignFrame.ControlsTreeEditingEnd(Sender: TObject; Node: TTreeNode;
+  Cancel: Boolean);
+var
+  S: String;
+  Sel: TComponent;
+begin
+  Sel := TComponent(ControlsTree.Selected.Data);
+  S := 'Rename ' + Sel.name + ' into ' + Node.Text;
+  Sel.name := Node.Text;
+  ModifiedOutsideObjectInspector;
+  RecordUndo(S); //It'd be good if we set "ItemIndex" to index of "name" field, but there doesn't seem to be an easy way to
+  Node.Text := ComponentCaption(Sel);
```

Comment:
As per my comments in "undo" PR, I would suggest to merge `ModifiedOutsideObjectInspector` with `RecordUndo`, so 

```
ModifiedOutsideObjectInspector;
RecordUndo(S);
```

-> would be rather 

```
ModifiedOutsideObjectInspector(S);
```

I would also suggest to rename local variable `S` -> `UndoComment` (more descriptive).

Reply:
> merge ModifiedOutsideObjectInspector with RecordUndo
That would be done as a separate PR, as we've discussed yesterday.

"S" fixed in 112b252dba6a348ee86456ee5ec06bd8383b68dc.